File: purplex/problems_app/validation_service.py
from typing import Dict, List, Any, Tuple, Optional
import json
import re
from django.core.exceptions import ValidationError
from .types import ValidationResult, ValidationError as CustomValidationError

class ProblemValidationService:
    """Comprehensive validation service for problems and test cases"""
    
    # Python keywords that cannot be used as function names
    PYTHON_KEYWORDS = {
        'and', 'as', 'assert', 'break', 'class', 'continue', 'def', 'del',
        'elif', 'else', 'except', 'exec', 'finally', 'for', 'from', 'global',
        'if', 'import', 'in', 'is', 'lambda', 'not', 'or', 'pass', 'print',
        'raise', 'return', 'try', 'while', 'with', 'yield', 'None', 'True', 'False'
    }
    
    def validate_problem_data(self, data: Dict[str, Any]) -> ValidationResult:
        """
        Validate complete problem data
        
        Args:
            data: Problem data dictionary to validate
            
        Returns:
            ValidationResult: Validation result with errors and warnings
        """
        errors = []
        warnings = []
        
        # Required field validations
        required_fields = ['title', 'description', 'function_name', 'reference_solution']
        for field in required_fields:
            if not data.get(field, '').strip():
                errors.append(CustomValidationError(
                    field=field,
                    message=f'{field.replace("_", " ").title()} is required',
                    code='REQUIRED'
                ))
        
        # Title validation
        if data.get('title'):
            title = data['title'].strip()
            if len(title) < 3:
                errors.append(CustomValidationError(
                    field='title',
                    message='Title must be at least 3 characters long',
                    code='MIN_LENGTH'
                ))
            elif len(title) > 200:
                errors.append(CustomValidationError(
                    field='title',
                    message='Title must not exceed 200 characters',
                    code='MAX_LENGTH'
                ))
        
        # Description validation
        if data.get('description'):
            description = data['description'].strip()
            if len(description) < 10:
                errors.append(CustomValidationError(
                    field='description',
                    message='Description must be at least 10 characters long',
                    code='MIN_LENGTH'
                ))
        
        # Function name validation
        if data.get('function_name'):
            function_name = data['function_name'].strip()
            validation_result = self.validate_function_name(function_name)
            if not validation_result[0]:
                errors.append(CustomValidationError(
                    field='function_name',
                    message=validation_result[1],
                    code='INVALID_FORMAT'
                ))
        
        # Reference solution validation
        if data.get('reference_solution'):
            solution = data['reference_solution'].strip()
            if len(solution) < 10:
                errors.append(CustomValidationError(
                    field='reference_solution',
                    message='Reference solution must be at least 10 characters long',
                    code='MIN_LENGTH'
                ))
            
            # Check if solution contains the function name
            function_name = data.get('function_name', '').strip()
            if function_name and f'def {function_name}' not in solution:
                warnings.append(CustomValidationError(
                    field='reference_solution',
                    message=f'Reference solution should define function "{function_name}"',
                    code='MISSING_FUNCTION_DEF'
                ))
        
        # Numeric field validations
        if 'memory_limit' in data:
            value = data['memory_limit']
            if not isinstance(value, int) or value < 32:
                errors.append(CustomValidationError(
                    field='memory_limit',
                    message='Memory limit must be at least 32 MB',
                    code='MIN_VALUE'
                ))
            elif value > 1024:
                errors.append(CustomValidationError(
                    field='memory_limit',
                    message='Memory limit must not exceed 1024 MB',
                    code='MAX_VALUE'
                ))
        
        # Test cases validation
        if 'test_cases' in data:
            test_cases = data['test_cases']
            if not isinstance(test_cases, list) or len(test_cases) == 0:
                errors.append(CustomValidationError(
                    field='test_cases',
                    message='At least one test case is required',
                    code='MIN_LENGTH'
                ))
            else:
                # Validate each test case
                for i, test_case in enumerate(test_cases):
                    tc_errors = self.validate_test_case_data(test_case)
                    for error in tc_errors:
                        errors.append(CustomValidationError(
                            field=f'test_cases.{i}.{error.field}',
                            message=error.message,
                            code=error.code
                        ))
                
                # Check for sample test cases
                has_sample = any(tc.get('is_sample', False) for tc in test_cases)
                if not has_sample:
                    warnings.append(CustomValidationError(
                        field='test_cases',
                        message='Consider marking at least one test case as a sample',
                        code='NO_SAMPLE_TESTS'
                    ))
        
        # Tags validation
        if 'tags' in data:
            tags = data['tags']
            if not isinstance(tags, list):
                errors.append(CustomValidationError(
                    field='tags',
                    message='Tags must be a list of strings',
                    code='INVALID_TYPE'
                ))
            else:
                for i, tag in enumerate(tags):
                    if not isinstance(tag, str):
                        errors.append(CustomValidationError(
                            field=f'tags.{i}',
                            message='Each tag must be a string',
                            code='INVALID_TYPE'
                        ))
                    elif len(tag.strip()) == 0:
                        errors.append(CustomValidationError(
                            field=f'tags.{i}',
                            message='Tags cannot be empty',
                            code='EMPTY_VALUE'
                        ))
        
        return ValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
            warnings=warnings
        )
    
    def validate_function_name(self, name: str) -> Tuple[bool, str]:
        """
        Validate Python function name
        
        Args:
            name: Function name to validate
            
        Returns:
            Tuple[bool, str]: (is_valid, error_message)
        """
        if not name:
            return False, "Function name cannot be empty"
        
        # Check Python identifier pattern
        if not re.match(r'^[a-zA-Z_][a-zA-Z0-9_]*$', name):
            return False, "Function name must be a valid Python identifier"
        
        # Check against Python keywords
        if name.lower() in self.PYTHON_KEYWORDS:
            return False, f"'{name}' is a Python keyword and cannot be used as a function name"
        
        # Check for common conventions
        if name.startswith('__') and name.endswith('__'):
            return False, "Dunder methods (names starting and ending with __) are not allowed"
        
        return True, ""
    
    def validate_test_case_data(self, test_case: Dict[str, Any]) -> List[CustomValidationError]:
        """
        Validate individual test case data
        
        Args:
            test_case: Test case data dictionary
            
        Returns:
            List[CustomValidationError]: List of validation errors
        """
        errors = []
        
        # Required fields
        if 'inputs' not in test_case:
            errors.append(CustomValidationError(
                field='inputs',
                message='Test case inputs are required',
                code='REQUIRED'
            ))
        else:
            inputs = test_case['inputs']
            if not isinstance(inputs, list):
                errors.append(CustomValidationError(
                    field='inputs',
                    message='Test case inputs must be a list',
                    code='INVALID_TYPE'
                ))
        
        if 'expected_output' not in test_case:
            errors.append(CustomValidationError(
                field='expected_output',
                message='Test case expected output is required',
                code='REQUIRED'
            ))
        else:
            # Validate that expected_output is JSON serializable
            try:
                json.dumps(test_case['expected_output'])
            except (TypeError, ValueError):
                errors.append(CustomValidationError(
                    field='expected_output',
                    message='Expected output must be JSON serializable',
                    code='INVALID_JSON'
                ))
        
        # Boolean field validation
        for field in ['is_hidden', 'is_sample']:
            if field in test_case and not isinstance(test_case[field], bool):
                errors.append(CustomValidationError(
                    field=field,
                    message=f'{field} must be a boolean value',
                    code='INVALID_TYPE'
                ))
        
        # Order validation
        if 'order' in test_case:
            order = test_case['order']
            if not isinstance(order, int) or order < 0:
                errors.append(CustomValidationError(
                    field='order',
                    message='Order must be a non-negative integer',
                    code='INVALID_VALUE'
                ))
        
        # Description validation
        if 'description' in test_case:
            description = test_case['description']
            if description and len(str(description).strip()) > 200:
                errors.append(CustomValidationError(
                    field='description',
                    message='Test case description must not exceed 200 characters',
                    code='MAX_LENGTH'
                ))
        
        return errors
    
    def validate_test_case_json(self, json_string: str) -> Tuple[bool, Any, Optional[str]]:
        """
        Validate JSON string for test case data
        
        Args:
            json_string: JSON string to validate
            
        Returns:
            Tuple[bool, Any, Optional[str]]: (is_valid, parsed_value, error_message)
        """
        try:
            parsed = json.loads(json_string)
            return True, parsed, None
        except json.JSONDecodeError as e:
            return False, None, f"Invalid JSON: {str(e)}"
        except Exception as e:
            return False, None, f"Unexpected error: {str(e)}"

File: purplex/client/src/services/validationService.ts
import {
  ProblemCreateRequest,
  TestCaseDisplay,
  TestCaseValidation,
  ValidationError
} from '../types';
import {
  autoDetectAndConvert,
  autoDetectTypeFromInput,
  getPlaceholderForType,
  parseTypeAnnotation,
  validateValueAgainstType
} from '../utils/typeSystem';

export interface ValidationResult {
  isValid: boolean;
  errors: ValidationError[];
  warnings: ValidationError[];
}

export interface JSONValidationResult {
  isValid: boolean;
  parsed?: unknown;
  error?: string;
}

class ValidationServiceImpl {
  /**
   * Validate complete problem data
   * @param problem - Partial problem data to validate
   * @returns Validation result with errors and warnings
   */
  validateProblem(problem: Partial<ProblemCreateRequest>): ValidationResult {
    const errors: ValidationError[] = [];
    const warnings: ValidationError[] = [];

    // Required field validations
    if (!problem.title?.trim()) {
      errors.push({
        field: 'title',
        message: 'Title is required',
        code: 'REQUIRED'
      });
    }

    if (!problem.function_name?.trim()) {
      errors.push({
        field: 'function_name',
        message: 'Function name is required',
        code: 'REQUIRED'
      });
    } else if (!this.validateFunctionName(problem.function_name)) {
      errors.push({
        field: 'function_name',
        message: 'Function name must be a valid Python identifier',
        code: 'INVALID_FORMAT'
      });
    }

    if (!problem.reference_solution?.trim()) {
      errors.push({
        field: 'reference_solution',
        message: 'Reference solution is required',
        code: 'REQUIRED'
      });
    }

    // Test case validations
    if (!problem.test_cases || problem.test_cases.length === 0) {
      errors.push({
        field: 'test_cases',
        message: 'At least one test case is required',
        code: 'MIN_LENGTH'
      });
    } else {
      // Validate each test case
      problem.test_cases.forEach((testCase, index) => {
        const tcValidation = this.validateTestCase({
          ...testCase,
          inputsString: JSON.stringify(testCase.inputs),
          expectedString: JSON.stringify(testCase.expected_output)
        } as TestCaseDisplay);

        if (!tcValidation.isValid) {
          tcValidation.errors.forEach(error => {
            errors.push({
              field: `test_cases.${index}`,
              message: error,
              code: 'INVALID_TEST_CASE'
            });
          });
        }
      });

      // Check for sample test cases
      const hasSampleTestCase = problem.test_cases.some(tc => tc.is_sample);
      if (!hasSampleTestCase) {
        warnings.push({
          field: 'test_cases',
          message: 'Consider marking at least one test case as a sample',
          code: 'NO_SAMPLE_TESTS'
        });
      }
    }

    // Numeric field validations
    if (problem.memory_limit !== undefined && problem.memory_limit < 32) {
      errors.push({
        field: 'memory_limit',
        message: 'Memory limit must be at least 32 MB',
        code: 'MIN_VALUE'
      });
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings
    };
  }

  /**
   * Validate a single test case with enhanced type checking
   * @param testCase - Test case to validate
   * @param functionSignature - Optional function signature for parameter validation
   * @returns Test case validation result
   */
  validateTestCase(testCase: TestCaseDisplay, functionSignature?: string): TestCaseValidation {
    const errors: string[] = [];
    const warnings: string[] = [];

    // Validate inputs JSON
    const inputsValidation = this.validateJSON(testCase.inputsString);
    if (!inputsValidation.isValid) {
      errors.push(`Invalid inputs JSON: ${inputsValidation.error}`);
    } else if (!Array.isArray(inputsValidation.parsed)) {
      errors.push('Inputs must be an array');
    } else if (functionSignature) {
      // Enhanced parameter type validation
      const parameterErrors = this.validateParameterTypes(inputsValidation.parsed, functionSignature);
      errors.push(...parameterErrors);
    }

    // Validate expected output JSON with type checking
    const outputValidation = this.validateJSON(testCase.expectedString);
    if (!outputValidation.isValid) {
      errors.push(`Invalid expected output JSON: ${outputValidation.error}`);
    } else if (functionSignature) {
      // Enhanced return type validation
      const returnTypeErrors = this.validateReturnType(outputValidation.parsed, functionSignature);
      errors.push(...returnTypeErrors);
    }

    // Check for description
    if (!testCase.description?.trim()) {
      warnings.push('Consider adding a description for this test case');
    }

    // Check for edge cases
    if (inputsValidation.isValid && Array.isArray(inputsValidation.parsed)) {
      if (inputsValidation.parsed.length === 0) {
        warnings.push('Test case has no input parameters');
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings
    };
  }

  /**
   * Validate parameter types against function signature
   * @param inputs - Array of input values
   * @param functionSignature - Function signature string
   * @returns Array of validation error messages
   */
  private validateParameterTypes(inputs: unknown[], functionSignature: string): string[] {
    const errors: string[] = [];
    
    try {
      const parameters = this.parseFunctionParameters(functionSignature);
      
      if (parameters.length !== inputs.length) {
        errors.push(`Expected ${parameters.length} parameters, got ${inputs.length}`);
        return errors;
      }

      inputs.forEach((input, index) => {
        if (index < parameters.length) {
          const param = parameters[index];
          const typeSpec = parseTypeAnnotation(param.type);
          const validationResult = validateValueAgainstType(input, typeSpec);
          
          if (!validationResult.valid) {
            const paramName = param.name || `parameter ${index + 1}`;
            let errorMsg = `${paramName}: ${validationResult.error}`;
            if (validationResult.suggestion) {
              errorMsg += ` (try: ${validationResult.suggestion})`;
            }
            errors.push(errorMsg);
          }
        }
      });
    } catch (error) {
      errors.push(`Failed to parse function signature: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }

    return errors;
  }

  /**
   * Validate return type against function signature
   * @param output - Expected output value
   * @param functionSignature - Function signature string
   * @returns Array of validation error messages
   */
  private validateReturnType(output: unknown, functionSignature: string): string[] {
    const errors: string[] = [];
    
    try {
      const returnType = this.parseFunctionReturnType(functionSignature);
      if (returnType && returnType !== 'Any') {
        const typeSpec = parseTypeAnnotation(returnType);
        const validationResult = validateValueAgainstType(output, typeSpec);
        
        if (!validationResult.valid) {
          let errorMsg = `Return value: ${validationResult.error}`;
          if (validationResult.suggestion) {
            errorMsg += ` (try: ${validationResult.suggestion})`;
          }
          errors.push(errorMsg);
        }
      }
    } catch (error) {
      errors.push(`Failed to parse return type: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }

    return errors;
  }

  /**
   * Parse function parameters from signature
   * @param functionSignature - Function signature string
   * @returns Array of parameter objects with name and type
   */
  private parseFunctionParameters(functionSignature: string): Array<{name: string; type: string}> {
    const regex = /def\s+\w+\s*\((.*?)\)/;
    const match = functionSignature.match(regex);
    
    if (!match) {
      return [];
    }
    
    const paramsStr = match[1].trim();
    if (!paramsStr) {
      return [];
    }
    
    const params: Array<{name: string; type: string}> = [];
    const paramRegex = /(\w+)\s*:\s*([^,]+)/g;
    let paramMatch;
    
    while ((paramMatch = paramRegex.exec(paramsStr)) !== null) {
      params.push({
        name: paramMatch[1],
        type: paramMatch[2].trim()
      });
    }
    
    return params;
  }

  /**
   * Parse return type from function signature
   * @param functionSignature - Function signature string
   * @returns Return type string or null if not specified
   */
  private parseFunctionReturnType(functionSignature: string): string | null {
    const regex = /def\s+\w+\s*\([^)]*\)\s*->\s*([^:]+):/;
    const match = functionSignature.match(regex);
    
    return match ? match[1].trim() : null;
  }

  /**
   * Validate JSON string with enhanced error reporting
   * @param jsonString - JSON string to validate
   * @returns JSON validation result with parsed value if valid
   */
  validateJSON(jsonString: string): JSONValidationResult {
    if (!jsonString.trim()) {
      return {
        isValid: false,
        error: 'Empty JSON string'
      };
    }

    try {
      const parsed = JSON.parse(jsonString);
      return {
        isValid: true,
        parsed
      };
    } catch (error) {
      let errorMessage = 'Invalid JSON';
      
      if (error instanceof SyntaxError) {
        // Provide more specific JSON syntax error messages
        const message = error.message.toLowerCase();
        if (message.includes('unexpected token')) {
          errorMessage = 'Invalid JSON syntax - check for missing quotes, commas, or brackets';
        } else if (message.includes('unexpected end')) {
          errorMessage = 'Incomplete JSON - missing closing bracket or quote';
        } else {
          errorMessage = `JSON syntax error: ${error.message}`;
        }
      }

      return {
        isValid: false,
        error: errorMessage
      };
    }
  }

  /**
   * Validate and suggest improvements for input values
   * @param inputValue - Raw input value as string
   * @param expectedType - Expected Python type annotation
   * @returns Validation result with suggestions
   */
  validateInputValue(inputValue: string, expectedType?: string): {
    isValid: boolean;
    parsed?: unknown;
    error?: string;
    suggestion?: string;
    confidence?: 'high' | 'medium' | 'low';
  } {
    if (!inputValue.trim()) {
      return {
        isValid: false,
        error: 'Empty input value',
        suggestion: expectedType ? getPlaceholderForType(expectedType) : 'Enter a value'
      };
    }

    // Auto-detect type from input
    const typeInfo = autoDetectTypeFromInput(inputValue);
    
    if (typeInfo.detected === 'invalid') {
      return {
        isValid: false,
        error: 'Invalid input format',
        suggestion: expectedType ? 
          `Expected ${expectedType}. Try: ${getPlaceholderForType(expectedType)}` :
          'Check your input syntax',
        confidence: typeInfo.confidence
      };
    }

    // Try to convert the value
    try {
      const converted = autoDetectAndConvert(inputValue);
      
      // If expected type is provided, validate against it
      if (expectedType) {
        const typeSpec = parseTypeAnnotation(expectedType);
        const validationResult = validateValueAgainstType(converted, typeSpec);
        
        if (!validationResult.valid) {
          return {
            isValid: false,
            parsed: converted,
            error: validationResult.error,
            suggestion: validationResult.suggestion,
            confidence: typeInfo.confidence
          };
        }
      }

      return {
        isValid: true,
        parsed: converted,
        confidence: typeInfo.confidence
      };
    } catch (error) {
      return {
        isValid: false,
        error: error instanceof Error ? error.message : 'Conversion failed',
        suggestion: expectedType ? getPlaceholderForType(expectedType) : 'Check your input format',
        confidence: typeInfo.confidence
      };
    }
  }

  /**
   * Validate Python function name
   * @param name - Function name to validate
   * @returns True if valid Python identifier
   */
  validateFunctionName(name: string): boolean {
    if (!name) {return false;}
    
    // Python identifier regex: starts with letter or underscore, 
    // followed by letters, digits, or underscores
    const pythonIdentifierRegex = /^[a-zA-Z_][a-zA-Z0-9_]*$/;
    
    // Check against Python keywords
    const pythonKeywords = [
      'and', 'as', 'assert', 'break', 'class', 'continue', 'def', 'del',
      'elif', 'else', 'except', 'exec', 'finally', 'for', 'from', 'global',
      'if', 'import', 'in', 'is', 'lambda', 'not', 'or', 'pass', 'print',
      'raise', 'return', 'try', 'while', 'with', 'yield'
    ];

    return pythonIdentifierRegex.test(name) && !pythonKeywords.includes(name.toLowerCase());
  }
}

// Export singleton instance
export const validationService = new ValidationServiceImpl();